File: ball/js/heap.js
class Heap {
    constructor(less) {
        this.heap = [];
        this.less = less;
    }

    add(e) {
        this.heap.push(e);
        let i = this.heap.length - 1;
        while (i > 0) {
            let j = (i % 2 == 0) ? (Math.floor(i / 2) - 1) : Math.floor(i / 2);
            if (this.less(this.heap[i], this.heap[j])) {
                this.swap(i, j);
                i = j;
            } else {
                break;
            }
        }
    }

    swap(i, j) {
        let temp = this.heap[i];
        this.heap[i] = this.heap[j];
        this.heap[j] = temp;
    }

    pop() {
        if (this.heap.length == 0) {
            return null;
        }

        let e = this.heap[0];
        this.heap[0] = this.heap[this.heap.length - 1];
        this.heap.pop();

        if (this.heap.length > 1) {
            let i = 0;
            while (i < this.heap.length) {
                let m = i*2 + 1;
                let n = i*2 + 2;
                if (m >= this.heap.length) {
                    break;
                }
                if (n >= this.heap.length) {
                    // console.log("heap length:" + this.heap.length + " m:" + m + " i:" + i);
                    // console.log("m:" + objToString(this.heap[m]));
                    // console.log("i:" + objToString(this.heap[i]));
                    if (this.less(this.heap[m], this.heap[i])) {
                        this.swap(m, i);
                    }
                    break;
                } else {
                    let k = this.less(this.heap[m], this.heap[n]) ? m : n;
                    if (this.less(this.heap[k], this.heap[i])) {
                        this.swap(k, i);
                        i = k;
                    } else {
                        break;
                    }
                }
            }
        }

        return e;
    }
 
    empty() {
        return this.heap.length == 0;
    }

    count() {
        return this.heap.length;
    }

    clear() {
        this.heap.length = 0;
    }

    show() {
        let s = "";
        for (let e of this.heap) {
            s += objToString(e) + " \n";
        }
        console.log(s);
    }
}

File: balllua/ball_logic/heap.lua
local Heap = {}
Heap.__index = Heap

function Heap:new(less)
    local self = {}
    setmetatable(self, Heap)
    self.array = {}
    self.less = less
    return self
end

function Heap:add(e)
    table.insert(self.array, e)
    local i = #self.array
    while i > 1 do
        local parent = math.floor((i - 1) / 2 + 1)
        if (i - 1) % 2 == 0 then
            parent = math.floor((i - 1) / 2)
        end
        if self.less(self.array[i], self.array[parent]) then
            self:swap(i, parent)
            i = parent
        else
            break
        end
    end
end

function Heap:swap(i, j)
    self.array[i], self.array[j] = self.array[j], self.array[i]
end

function Heap:pop()
    if self:empty() then
        return
    end

    local size = #self.array
    local e = self.array[1]
    self.array[1] = self.array[size]
    self.array[size] = nil

    size = #self.array
    if size > 1 then
        local i = 1
        while i < size do
            local left = (i - 1) * 2 + 2
		    local right = (i - 1) * 2 + 3
            if left > size then
                break
            end

            if right > size then
                if self.less(self.array[left], self.array[i]) then
                    self:swap(left, i)
                end
                break
            else
                local j = left
                if self.less(self.array[right], self.array[left]) then
                    j = right
                end
                if self.less(self.array[j], self.array[i]) then
                    self:swap(i, j)
                    i = j
                else
                    break
                end
            end
        end
    end
    return e
end

function Heap:empty()
    return #self.array == 0
end

function Heap:count()
    return #self.array
end

function Heap:clear()
    self.array = {}
end

return Heap
